src/language_features/semantic_tokens.rs

```diff
@@ -65,18 +66,53 @@ pub fn tokens_response(meta: EditorMeta, tokens: SemanticTokensResult, ctx: &mut
                     end: Position::new(line, start + length),
                 };
                 let range = lsp_range_to_kakoune(&range, &document.text, ctx.offset_encoding);
-                let token = &legend.token_types[token_type as usize];
-                (0..32)
+                let token_name = legend.token_types[token_type as usize].as_str();
+                let token_modifiers: Vec<&SemanticTokenModifier> = (0..32)
+                    // Find bits in the mask that equal `1`
                     .filter(|bit| ((token_modifiers_bitset >> bit) & 1u32) == 1u32)
+                    // Map bits to modifiers
                     .map(|bit| &legend.token_modifiers[bit as usize])
-                    .filter_map(|token| ctx.config.semantic_token_modifiers.get(token.as_str()))
-                    .chain(ctx.config.semantic_tokens.get(token.as_str()))
-                    .next()
-                    .map(|face| format!("{}|{}", range, face))
+                    .collect();
+
+                let mut candidates = ctx
+                    .config
+                    .semantic_tokens
+                    .iter()
+                    .filter_map(|token_config| {
+                        if token_name != token_config.name {
+                            return None;
+                        }
+
+                        // All the config's modifiers must exist on the token for this
+                        // config to match.
+                        if !token_config
+                            .modifiers
+                            .iter()
+                            .all(|modifier| token_modifiers.contains(&modifier))
+                        {
+                            return None;
+                        }
+
+                        // But not all the token's modifiers must exist on the config.
+                        // Therefore, we get a count of matching ones and sort by that.
+                        let modifier_count = token_modifiers
+                            .iter()
+                            .filter(|modifier| token_config.modifiers.contains(modifier))
+                            .count();
+
+                        Some((&token_config.face, modifier_count))
+                    })
+                    .collect::<Vec<_>>();
+
+                // Sort by number of matching modifiers, in descending order
+                candidates.sort_by_key(|val| val.1);
+                candidates.reverse();
+                candidates.first().map(|val| format!("{}|{}", range, val.0))
```

Comment:
It sucks that we can't do a descending sort by simply negating `val.1`. We could use bitwise negation but that's hardly intuitive.

I guess we can use `last()` instead of `first()`